providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+                }
+            }
+            for (l = ctx->lanes - ctx->threads; l < ctx->lanes; ++l) {
+                if (crypto_thread_join(ctx->openssl_ctx, t[l], NULL) == 0)
+                    goto fail;
+                t[l] = NULL;
+            }
+        }
+    }
+
+    crypto_thread_clean(ctx->openssl_ctx, NULL);
+
+    OPENSSL_free(t_data);
+    OPENSSL_free(t);
+
+    return 1;
+
+fail:
+    if (t_data != NULL)
+        OPENSSL_free(t_data);
+    if (t != NULL)
+        OPENSSL_free(t);
+    return 0;
+}
+
+#endif /* ARGON2_NO_THREADS OPENSSL_THREADS */
+
+static int fill_mem_blocks_st(ARGON2_CTX *ctx)
+{
+    uint32_t r, s, l;
+
+    for (r = 0; r < ctx->passes; ++r)
+        for (s = 0; s < ARGON2_SYNC_POINTS; ++s)
+            for (l = 0; l < ctx->lanes; ++l)
+                fill_segment(ctx, r, l, s, 0);
+    return 1;
+}
+
+static ossl_inline int fill_memory_blocks(ARGON2_CTX *ctx)
+{
+    return ctx->threads == 1 ? fill_mem_blocks_st(ctx) : fill_mem_blocks_mt(ctx);
+}
+
+static int validate_inputs(const ARGON2_CTX *ctx)
+{
+    /*
+     * due to -Werror=type-limits, some of the comparisons are made as
+     * X+1 > Y+1 rather than X > Y, as this caused problems when some
+     * lower limits were set to zero.
+     */
+
+    if (ctx == NULL)
+        return 0;
+
+    switch (ctx->type) {
+    case ARGON2_D:
+    case ARGON2_I:
+    case ARGON2_ID:
+        break;
+    default:
+        return 0;
+    }
+
+    if (ctx->out == NULL)
+        return 0;
+    if (ctx->outlen < ARGON2_MIN_OUTLEN)
+        return 0;
+    if (ctx->outlen > ARGON2_MAX_OUTLEN)
+        return 0;
+    if (ctx->pwd == NULL && ctx->pwdlen != 0)
+        return 0;
+    if (ctx->pwdlen+1 < ARGON2_MIN_PWD_LENGTH+1)
+        return 0;
+    if (ctx->pwdlen > ARGON2_MAX_PWD_LENGTH)
+        return 0;
+    if (ctx->salt == NULL && ctx->saltlen != 0)
+        return 0;
+    if (ctx->saltlen < ARGON2_MIN_SALT_LENGTH)
+        return 0;
+    if (ctx->saltlen > ARGON2_MAX_SALT_LENGTH)
+        return 0;
+    if (ctx->secret == NULL && ctx->secretlen != 0)
+        return 0;
+    if (ctx->secret != NULL && ARGON2_MIN_SECRET+1 > ctx->secretlen+1)
+        return 0;
+    if (ctx->secret != NULL && ARGON2_MAX_SECRET < ctx->secretlen)
+        return 0;
+    if (ctx->ad == NULL && ctx->adlen != 0)
+        return 0;
+    if (ctx->ad != NULL && ARGON2_MIN_AD_LENGTH+1 > ctx->adlen+1)
+        return 0;
+    if (ctx->ad != NULL && ARGON2_MAX_AD_LENGTH < ctx->adlen)
+        return 0;
+    if (ctx->m_cost < ARGON2_MIN_MEMORY)
+        return 0;
+    if (ctx->m_cost >= (uint64_t) ARGON2_MAX_MEMORY)
+        if (ctx->m_cost != (uint64_t) ARGON2_MAX_MEMORY)
+            return 0;
+    if (ctx->m_cost < 8 * ctx->lanes)
+        return 0;
+    if (ctx->t_cost < ARGON2_MIN_TIME)
+        return 0;
+    if (ctx->t_cost > ARGON2_MAX_TIME)
+        return 0;
+    if (ctx->lanes < ARGON2_MIN_LANES)
+        return 0;
+    if (ctx->lanes > ARGON2_MAX_LANES)
+        return 0;
```

Comment:
collapsing this to just do a single return 0 would be cleaner.

Reply:
not thrilled with the null-pointer check.  if you give a size for something, and a NULL pointer, you did something wrong and a segfault seems reasonable and the best way to find and fix the error.

Reply:
An assert would be better than segfaulting.